### PR TITLE
Add blood level widget for XE vampire characters

### PR DESCRIPTION
**Pull request by Maleclypse**
#### Summary
Mods "Blood level widget for Xedra Evolved Vampire PCs"

#### Purpose of change
Add a widget for blood levels.

#### Describe the solution
I'm cargo culting.

#### Describe alternatives you've considered
Bribing someone smarter.

#### Testing
It no work yet.

### CODE REVIEW

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    "id": "vampire_blood_levels",
+    "type": "widget",
+    "label": "Blood",
+    "style": "text",
+    "clauses": [
+      {
+        "id": "slumber",
+        "text": "Eternal Slumber",
+        "color": "dark_gray_red",
+        "condition": { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": -11999 } ] }
```
Comment (Maleclypse) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:12]:
```suggestion
        "condition": { "compare_num": [ { "vitamin": "human_blood_vitamin" }, "<=", { "const": -11999 } ] }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        "id": "slumber",
+        "text": "Eternal Slumber",
+        "color": "dark_gray_red",
+        "condition": { "compare_num": [ { "vitamin": "human_blood_vitamin" }, "<=", { "const": -11999 } ] }
+      },
+      {
+        "id": "near_death",
+        "text": "Corpse-like",
+        "color": "light_gray",
+        "condition": { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": -6401 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:18]:
```suggestion
        "condition": {
          "and": [
            { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": -6401 } ] },
            { "compare_num": [ { "vitamin": "human_blood_vitamin" }, ">", { "const": -11999 } ] }
          ]
        }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      },
+      {
+        "id": "starving",
+        "text": "Hypomia",
+        "color": "yellow",
+        "condition": { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": -5601 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:24]:
```suggestion
        "condition": {
          "and": [
            { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": -5601 } ] },
            { "compare_num": [ { "vitamin": "human_blood_vitamin" }, ">", { "const": -6401 } ] }
          ]
        }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      },
+      {
+        "id": "hungry",
+        "text": "Sallow",
+        "color": "light_green",
+        "condition": { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": -4800 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:30]:
```suggestion
        "condition": {
          "and": [
            { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": -4800 } ] },
            { "compare_num": [ { "vitamin": "human_blood_vitamin" }, ">", { "const": -5601 } ] }
          ]
        }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      },
+      {
+        "id": "content",
+        "text": "Content",
+        "color": "green",
+        "condition": { "compare_num": [ { "u_val": "human_blood_vitamin" }, ">=", { "const": 0 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:36]:
```suggestion
        "condition": {
          "and": [
            { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": 1800 } ] },
            { "compare_num": [ { "vitamin": "human_blood_vitamin" }, ">", { "const": -4800 } ] }
          ]
        }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      },
+      {
+        "id": "flush",
+        "text": "FLUSH",
+        "color": "light_red",
+        "condition": { "compare_num": [ { "u_val": "human_blood_vitamin" }, ">=", { "const": 1800 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:42]:
```suggestion
        "condition": {
          "and": [
            { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": 2401 } ] },
            { "compare_num": [ { "vitamin": "human_blood_vitamin" }, ">", { "const": 1800 } ] }
          ]
        }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      },
+      {
+        "id": "engorged",
+        "text": "Engorged",
+        "color": "red",
+        "condition": { "compare_num": [ { "u_val": "human_blood_vitamin" }, ">=", { "const": 2401 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:48]:
```suggestion
        "condition": {
          "and": [
            { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": 4401 } ] },
            { "compare_num": [ { "vitamin": "human_blood_vitamin" }, ">", { "const": 2401 } ] }
          ]
        }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      },
+      {
+        "id": "swollen",
+        "text": "Tick-like",
+        "color": "red_cyan",
+        "condition": { "compare_num": [ { "u_val": "human_blood_vitamin" }, ">=", { "const": 4401 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:54]:
```suggestion
        "condition": { "compare_num": [ { "u_val": "human_blood_vitamin" }, ">", { "const": 4401 } ] }
```

**Review by dseguin — COMMENTED** [review]
It won't work because you need to add upper and lower bounds to your vitamin levels checks
(i.e.: more than 1800 `and` less than 2401 -> "FLUSH")

**Comment by Maleclypse** [comment]
@dseguin 
Here's the error I'm getting
(all_mods)=> Backtrace emission took 1 seconds.
(all_mods)=> (continued from above) ERROR : Terminated:
Error: (all_mods)=> unrecognized number source in { "vitamin": "human_blood_vitamin" }

        "text": "Eternal Slumber",
        "color": "dark_gray_red",
        "condition": { "compare_num": [
                                       ^
                                        { "vitamin": "human_blood_vitamin" }, "<=", { "const": -11999 } ] }
      },
      {

(all_mods)=> 13:44:25.[549](https://github.com/CleverRaven/Cataclysm-DDA/actions/runs/4443735108/jobs/7801912068?pr=64312#step:15:550) INFO : Make sure that you're in the correct working directory and your data isn't corrupted.

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    "id": "vampire_blood_levels",
+    "type": "widget",
+    "label": "Blood",
+    "style": "text",
+    "clauses": [
+      {
+        "id": "slumber",
+        "text": "Eternal Slumber",
+        "color": "dark_gray_red",
+        "condition": { "compare_num": [ { "vitamin": "human_blood_vitamin" }, "<=", { "const": -11999 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:12]:
```suggestion
        "condition": { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": -11999 } ] }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+      },
+      {
+        "id": "near_death",
+        "text": "Corpse-like",
+        "color": "light_gray",
+        "condition": {
+          "and": [
+            { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": -6401 } ] },
+            { "compare_num": [ { "vitamin": "human_blood_vitamin" }, ">", { "const": -11999 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:21]:
```suggestion
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": -6401 } ] },
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, ">", { "const": -11999 } ] }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        }
+      },
+      {
+        "id": "starving",
+        "text": "Hypomia",
+        "color": "yellow",
+        "condition": {
+          "and": [
+            { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": -5601 } ] },
+            { "compare_num": [ { "vitamin": "human_blood_vitamin" }, ">", { "const": -6401 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:32]:
```suggestion
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": -5601 } ] },
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, ">", { "const": -6401 } ] }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        }
+      },
+      {
+        "id": "hungry",
+        "text": "Sallow",
+        "color": "light_green",
+        "condition": {
+          "and": [
+            { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": -4800 } ] },
+            { "compare_num": [ { "vitamin": "human_blood_vitamin" }, ">", { "const": -5601 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:43]:
```suggestion
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": -4800 } ] },
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, ">", { "const": -5601 } ] }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        }
+      },
+      {
+        "id": "content",
+        "text": "Content",
+        "color": "green",
+        "condition": {
+          "and": [
+            { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": 1800 } ] },
+            { "compare_num": [ { "vitamin": "human_blood_vitamin" }, ">", { "const": -4800 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:54]:
```suggestion
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": 1800 } ] },
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, ">", { "const": -4800 } ] }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        }
+      },
+      {
+        "id": "flush",
+        "text": "FLUSH",
+        "color": "light_red",
+        "condition": {
+          "and": [
+            { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": 2401 } ] },
+            { "compare_num": [ { "vitamin": "human_blood_vitamin" }, ">", { "const": 1800 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:65]:
```suggestion
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": 2401 } ] },
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, ">", { "const": 1800 } ] }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        }
+      },
+      {
+        "id": "engorged",
+        "text": "Engorged",
+        "color": "red",
+        "condition": {
+          "and": [
+            { "compare_num": [ { "u_val": "human_blood_vitamin" }, "<=", { "const": 4401 } ] },
+            { "compare_num": [ { "vitamin": "human_blood_vitamin" }, ">", { "const": 2401 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:76]:
```suggestion
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": 4401 } ] },
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, ">", { "const": 2401 } ] }
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+          ]
+        }
+      },
+      {
+        "id": "swollen",
+        "text": "Tick-like",
+        "color": "red_cyan",
+        "condition": { "compare_num": [ { "u_val": "human_blood_vitamin" }, ">", { "const": 4401 } ] }
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:84]:
```suggestion
        "condition": { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, ">", { "const": 4401 } ] }
```

**Review by dseguin — COMMENTED** [review]
This should fix it :+1:

**Comment by Maleclypse** [comment]
Oh wow, I feel silly that I missed that. Thanks!

**Comment by GuardianDll** [comment]
Making a separate sidebar only for vampire text is not good approach imo
I'll try to investigate it futher

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+          ]
+        }
+      },
+      {
+        "id": "hungry",
+        "text": "Sallow",
+        "color": "light_green",
+        "condition": {
+          "and": [
+            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": -4800 } ] },
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:42]:
```suggestion
            { "u_has_effect": "vampire_virus" },
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": -4800 } ] },
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+          ]
+        }
+      },
+      {
+        "id": "content",
+        "text": "Content",
+        "color": "green",
+        "condition": {
+          "and": [
+            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": 1800 } ] },
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:53]:
```suggestion
            { "u_has_effect": "vampire_virus" },
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": 1800 } ] },
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+          ]
+        }
+      },
+      {
+        "id": "flush",
+        "text": "FLUSH",
+        "color": "light_red",
+        "condition": {
+          "and": [
+            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": 2401 } ] },
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:64]:
```suggestion
            { "u_has_effect": "vampire_virus" },
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": 2401 } ] },
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+          ]
+        }
+      },
+      {
+        "id": "engorged",
+        "text": "Engorged",
+        "color": "red",
+        "condition": {
+          "and": [
+            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": 4401 } ] },
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:75]:
```suggestion
            { "u_has_effect": "vampire_virus" },
            { "compare_num": [ { "u_val": "vitamin", "name": "human_blood_vitamin" }, "<=", { "const": 4401 } ] },
```

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json [hunk 21]
```diff
@@ -0,0 +1,88 @@
+[
+  {
+    "id": "vampire_blood_levels",
+    "type": "widget",
+    "label": "Blood",
+    "style": "text",
```
Comment (dseguin) [inline data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:6]:
```suggestion
    "style": "text",
    "flags": [ "W_DISABLED_WHEN_EMPTY" ],
```

**Review by dseguin — COMMENTED** [review]
You don't need to create a whole new sidebar, just add a condition to check if the player has the vampire virus effect. If they don't then the widget will be empty, and adding the W_DISABLED_WHEN_EMPTY flag should disable it in that case

File: data/mods/Xedra_Evolved/ui/vamp_blood_levels.json (hunk 21 above)
Comment (Maleclypse) [reply data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:6]: How does that work with negative numbers? Will it be disabled when negative?

Comment (dseguin) [reply data/mods/Xedra_Evolved/ui/vamp_blood_levels.json:6]: It means the widget is disabled only if there's nothing to display, i.e. when none of the clauses are matched. So if the player has no `vampire_virus` effect, none of the clauses here will match and the widget will be disabled

**Pull request merged**